types/statuses/index.d.ts

```diff
@@ -3,13 +3,36 @@
 // Definitions by: Tanguy Krotoff <https://github.com/tkrotoff>
 //                 BendingBender <https://github.com/BendingBender>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.7
+
+type NonNumericAscii =
+    '\x00' | '\x01' | '\x02' | '\x03' | '\x04' | '\x05' | '\x06' | '\x07' | '\b' | '\t' | '\n' | '\v' | '\f' | '\r' | '\x0E' | '\x0F'
+    | '\x10' | '\x11' | '\x12' | '\x13' | '\x14' | '\x15' | '\x16' | '\x17' | '\x18' | '\x19' | '\x1A' | '\x1B' | '\x1C' | '\x1D' | '\x1E' | '\x1F'
+    | ' ' | '!' | '"' | '#' | '$' | '%' | '&' | '\'' | '(' | ')' | '*' | '+' | ',' | '-' | '.' | '/'
+    | /* digits would be here */ ':' | ';' | '<' | '=' | '>' | '?'
+    | '@' | 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G' | 'H' | 'I' | 'J' | 'K' | 'L' | 'M' | 'N' | 'O'
+    | 'P' | 'Q' | 'R' | 'S' | 'T' | 'U' | 'V' | 'W' | 'X' | 'Y' | 'Z' | '[' | '\\' | ']' | '^' | '_'
+    | '`' | 'a' | 'b' | 'c' | 'd' | 'e' | 'f' | 'g' | 'h' | 'i' | 'j' | 'k' | 'l' | 'm' | 'n' | 'o'
+    | 'p' | 'q' | 'r' | 's' | 't' | 'u' | 'v' | 'w' | 'x' | 'y' | 'z' | '{' | '|' | '}' | '~' | '\x7F';
+
+type IsNumericString<S extends string> = S extends `${number}` ? any : never;
+
+type IsNonNumericString<S extends string> = S extends `${NonNumericAscii}${infer _}` ? any : never;
+
+type Result<S extends string> =
+    S extends IsNumericString<S>
+    ? string
+    : S extends IsNonNumericString<S>
+    ? number
+    : string|number;
```

Comment:
This type should be exposed for ease of use. It can be done by introducing a `namespace`. But it'll only work, if the namespace, the interface `Status`, and the constant `status` have the same name, like this:

```ts
export = status;
declare const status: status;
interface status {
    (code: number): string;
    <S extends string>(code: S): status.Result<S>;

    ...
}
declare namespace status {
    type Result<S extends string> = S extends IsNumericString<S>
        ? string
        : S extends IsNonNumericString<S>
        ? number
        : string | number;
}
```

And you can add a test for it:
```ts
type R = status.Result<any>;
```